### ISSUE TITLE
BUG REPORT [5.0pre3] - GUI cuts some articulation lists off towards the bottom

### ISSUE THREAD

**Issue by seventh-sam** [jtackaberry/reaticulate#155]
@jtackaberry 

One of two bugs I found testing 5.0pre3.  Will post the other shortly.

EDIT: Running REAPER v6.45, Windows 10 (Up to date)

Some articulation lists showed up in the GUI with an arbitrary part of the list cut off towards the bottom (see included screenshots).

I tested the following:
- Zooming in and out didn't fix anything; the cut-off just zooms with the rest of it.
- Switched the REAPER theme.  No effect.
- Tried several test banks, user banks, factory banks etc.  I thought there might be a common variable - such as lists with groups, or lists that have spacers - but no such luck.  It *seems* completely arbitrary.

![image](https://user-images.githubusercontent.com/70399072/149887662-b0e07403-b196-4a69-b0ff-a479d6cf25e5.png)
![image](https://user-images.githubusercontent.com/70399072/149887736-b8993914-2910-48cf-8ef0-f38cddc8eadf.png)
![image](https://user-images.githubusercontent.com/70399072/149887787-fb96d9d4-c383-4fea-bce1-786c47336132.png)

Let me know if there's anything else you need from me regarding this.

**Comment by jtackaberry** [comment]
Hm, this is a concerning one. I've not run into anything like this on my end.  A few questions:

1. What's the resolution of your Reaper window? 
2. What is your system UI scaling percentage set to in Windows? 
3. Does it go away if you resize the Reaper window smaller? 
4. Does this reproduce or you undock Reaticulate? At all undocked window sizes?

Thanks!

**Comment by seventh-sam** [comment]
@jtackaberry 

Okay, I've dug into this and narrowed down the culprit (which can hopefully lead to the solution!).

I was wrong before.  I went through all your questions with a fine tooth comb, tested all of those factors in combination with each other, and found the _only_ factor that causes this glitch seems to be the presence of //! spacer="" in the bank.  Any bank that had no spacers in it worked as it should.  I tested this with about two dozen of my own banks (with spacers in them) as well as test banks made post-installation of 5.0pre3 just to be sure, so I'm reasonably confident that something about how the spacers work is what's causing the glitch.

Furthermore:

- What threw me off initially is that it's not the presence of spacers _themselves_ in the bank that causes the glitch.  Rather, it's if there are a certain number of spacers in there in certain locations.  For instance, a bank with one spacer in a certain place might work perfectly fine, but if I move that spacer somewhere else in the bank or add more spacers, it _might_ glitch out.  I couldn't figure out a pattern to this, but since you know the code inside out I have a feeling you'll be able to.  I've included screenshots and GIFs below so you can see what I'm talking about.
- If there are multiple banks assigned, a _different_ glitch is introduced in which the banks "overlap" each other IF the first bank is glitched out to begin with.  I think this might illuminate the source of this glitch (more on that at the end of this post)  If the glitchy bank is the second, the banks do not overlap but the second bank still cuts off at a (seemingly) arbitrary point.  I've included screenshots of this as well.

_(Note: I tested the following behaviors with many banks both new and old, but I've only included screen captures of one for brevity's sake.  In other words, I'm able to reproduce the following regardless of monitor, resolution, date of bank creation, etc.)_

**Behavior if the bank has the "wrong" amount of spacers:**

![singlebank_glitchedspacers](https://user-images.githubusercontent.com/70399072/150031313-e7ad7dbb-819e-470f-93c0-a079185d02af.gif)

**Behavior if the bank has the "right" amount of spacers:**

![singlebank_rightspacers](https://user-images.githubusercontent.com/70399072/150031687-31ca5a5b-8359-449c-82ef-699445961def.gif)

**If multiple banks are loaded, and the glitched one is first:**

One column:
![image](https://user-images.githubusercontent.com/70399072/150032056-0beaf2f3-34a8-430a-b082-77d83b06b9de.png)

Two columns:
![image](https://user-images.githubusercontent.com/70399072/150032122-e1a173f7-eb83-4123-aca9-b3ff0f0ec4d8.png)

Three columns:
![image](https://user-images.githubusercontent.com/70399072/150032229-b83fda74-1f28-4455-bbcd-646a58b489bb.png)

_The above behaviors remain true regardless of docking or zoom size_

**If multiple banks are loaded, and the glitched one is second:**

![image](https://user-images.githubusercontent.com/70399072/150032428-1862b3b6-ef71-4fa0-b7ff-256597614b99.png)

----------

I can't help but notice that the locations on the GUI where the banks glitch out _just so happen_ to coincide with where the banks overlap if multiple banks are loaded up.  It looks to me like the presence of a certain amount of spacers is somehow screwing with the bottom boundaries that the GUI is drawing in anticipation of multiple banks being loaded up.  Think this is the culprit?

I hope all that helps!  Let me know if you need any clarification or further testing and I'd be more than happy to help.

**Comment by jtackaberry** [comment]
Wow, this is some brilliant sleuthing Sam. I wish all my bug reports looked like this. :)

Thanks to your excellent detail and inferences, I've been able to quickly reproduce the problem.  It's specifically related to the "[flow box](https://reapertoolkit.dev/class/rtk.FlowBox.html)" widget I wrote for rtk to do the adaptive multi-column layout and it's almost certainly a stupid oversight where I forgot to account for the top-padding (which is all the spacer is) in one of the calculations.

I had all kinds of fears that the issue was lower level than that but I'm now pretty sure it's an easy fix.

Thanks!
